### PR TITLE
Remove shared ref to underlying array of JSONFrameReader's Read arg.

### PR DESCRIPTION
**Pull request by benluddy**
#### What type of PR is this?

/kind bug
/sig api-machinery

#### What this PR does / why we need it:

When JSONFrameReader read a JSON object longer than the length of the destination slice, but not larger than the capacity of the destination slice, it would retain a reference to a slice sharing the same underlying array as the destination slice. If the underlying array is modified between calls to Read, corrupt frame data could be returned.

This is also called out in the io.Reader contract: "Implementations must not retain p."

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

The current usage in `k8s.io/apimachinery/pkg/runtime/serializer/streaming` does not appear to be affected. Even though it passes the frame reader slices with cap > len, it always either creates a new slice or ignores the slice contents on io.ErrShortBuffer.

#### Does this PR introduce a user-facing change?

```release-note
Fixed a bug in the JSON frame reader that could cause it to retain a reference to the underlying array of the byte slice passed to Read.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/framer/framer_test.go [hunk 1]
```diff
@@ -173,3 +173,17 @@ func TestJSONFrameReaderShortBuffer(t *testing.T) {
 		t.Fatalf("unexpected: %v %d %q", err, n, buf)
 	}
 }
+
+func TestJSONFrameReaderShortBufferNoUnderlyingArrayReuse(t *testing.T) {
+	b := bytes.NewBufferString("{}")
+	r := NewJSONFramedReader(io.NopCloser(b))
+	buf := make([]byte, 1, 2) // cap(buf) > len(buf) && cap(buf) <= len("{}")
+
+	if n, err := r.Read(buf); err != io.ErrShortBuffer || n != 1 || string(buf[:n]) != "{" {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/util/framer/framer_test.go:182]: This seems to fail the linter:
```
staging/src/k8s.io/apimachinery/pkg/util/framer/framer_test.go:182:28: comparing with != will fail on wrapped errors. Use errors.Is to check for a specific error (errorlint)
	if n, err := r.Read(buf); err != io.ErrShortBuffer || n != 1 || string(buf[:n]) != "{" {
```

**Comment by wojtek-t** [comment]
One minor comment - but it looks good overall.

/approve

File: staging/src/k8s.io/apimachinery/pkg/util/framer/framer_test.go (hunk 1 above)
Comment (benluddy) [reply staging/src/k8s.io/apimachinery/pkg/util/framer/framer_test.go]: Fixed, thanks.

**Comment by benluddy** [comment]
/test pull-kubernetes-e2e-gce

https://github.com/kubernetes/kubernetes/issues/93215#issuecomment-2103072871

**Pull request merged**